### PR TITLE
Migrate husky to lefthook

### PR DESCRIPTION
**Pull request by diegohaz**
## Motivation

The repository needed two dependencies for one Git hook. husky installed the hook and lint-staged ran the tasks over the staged files. The hook was also spread across three places: the `prepare` script, `.husky/pre-commit`, and a `lint-staged` field in `package.json`.

Lefthook covers both roles with one dependency, and it declares the hook, its file filters, and its task order in a single `lefthook.yml`.

## Solution

Replace `husky` and `lint-staged` with `lefthook`, and move the pre-commit tasks to `lefthook.yml`.

Before, in `package.json`:

```json
"lint-staged": {
  "package.json": "pnpm -r run --if-present clean",
  "*.{js,ts,tsx,md,mdx,json}": [
    "pnpm run sync-astro-types",
    "oxlint --disable-nested-config --fix --no-error-on-unmatched-pattern",
    "oxfmt --no-error-on-unmatched-pattern"
  ]
}
```

After, in `lefthook.yml`:

```yml
glob_matcher: doublestar

pre-commit:
  piped: true
  jobs:
    - name: clean
      glob: "**/package.json"
      run: pnpm -r run --if-present clean
      stage_fixed: true

    - name: lint
      glob: "**/*.{js,ts,tsx,md,mdx,json}"
      group:
        piped: true
        jobs:
          - name: sync-astro-types
            run: pnpm run sync-astro-types
          - name: oxlint
            run: pnpm exec oxlint --disable-nested-config --fix --no-error-on-unmatched-pattern {staged_files}
            stage_fixed: true
          - name: oxfmt
            run: pnpm exec oxfmt --no-error-on-unmatched-pattern {staged_files}
            stage_fixed: true
```

The same tools run over the same files in the same order. `stage_fixed` replaces the automatic `git add` that lint-staged performed after each task, and `piped` replaces its stop-on-first-failure behavior, which keeps `oxlint --fix` from rewriting source after `sync-astro-types` fails.

Two details are not obvious from the config:

- `glob_matcher: doublestar` selects Bash-like globs. Lefthook's default `gobwas` matcher resolves `**/package.json` against nested files only, so the file at the repository root would be skipped, and the job would need a second pattern. The explicit setting also keeps the patterns correct if the default ever changes.
- Lefthook does not add `node_modules/.bin` to `PATH`, so `oxlint` and `oxfmt` run through `pnpm exec`. Without it the hook fails with `oxlint: command not found`.

The hook is also slightly stricter than before. lint-staged ran the `package.json` group and the `*.{js,...}` group concurrently, so `oxfmt` could format a `package.json` before `clean` rewrote its exports. The hook level `piped: true` removes that race.

### Installing the hook

This follows lefthook's installation guide for npm. Lefthook installs the hooks from its own postinstall script, so there is no `prepare` script, and `pnpm-workspace.yaml` allows that build:

```yml
allowBuilds:
  lefthook: true
```

The entry is required either way: pnpm 11 fails the install with `ERR_PNPM_IGNORED_BUILDS` when a dependency has a build script and no entry.

That postinstall returns early when `CI` is set and `LEFTHOOK` is not, so CI never installs the hooks. It also does not fail when the working directory has no `.git`, which `ariakit react18` produces when it installs a copy of the workspace, although it does print a `fatal: not a git repository` block there where husky was silent.

Checkouts that ran the husky setup still point `core.hooksPath` at `.husky/_`, and lefthook writes its hook there, so hooks keep working without a manual step. Run this once anyway:

```bash
git config --unset-all --local core.hooksPath && rm -rf .husky
```

husky wrote that option as a relative path, which Git resolves per working directory. Until it is cleared, a linked worktree gets its own untracked `.husky/_/pre-commit` holding a machine specific path, and the main checkout keeps running husky's shim, which now resolves to a deleted file and silently does nothing. Clearing the option puts the hook in the default directory for every worktree.

### Other changes

- CI disables the hook with `LEFTHOOK=0` instead of `HUSKY=0` in the setup action. The three step level copies in `perf.yml` are gone. Two sat on baseline install steps in jobs that run the setup action earlier, and the third sat on the setup action step itself, where the composite exports the variable before its own install, so every one of them was already redundant. The husky specific note about `core.hooksPath` being rewritten under a temporary worktree stops applying once base commits carry lefthook. Until then the baseline install runs the base commit's husky `prepare`, which writes that option into the shared config. No perf job runs a hook or commits, so it has no effect.
- `ariakit clean` no longer declares the ignored `[files...]` argument. It existed only because lint-staged appended matched filenames to package scripts, and the `clean` job does not pass files.
- The reviewed root dependency set in `packages/ariakit-scripts/src/ci.ts` and the Renovate automerge group replace their `husky` and `lint-staged` entries with `lefthook`.
- Comments in `oxlint.config.ts` and `packages/ariakit-scripts/src/build.ts` refer to the pre-commit hook instead of lint-staged.
- `lefthook.yml` sits at the repository root, so a change to it alone now selects full CI, where `.husky/pre-commit` did not. That path fails closed and hook-only commits are rare.

### Known behavior changes

When a job fails, lefthook aborts the commit but keeps what the earlier jobs already fixed and re-staged. lint-staged reverted every task's changes instead. Unstaged hunks are still restored either way, so no work is lost, and the auto-fixes stay applied while you correct the remaining error.

husky always set `core.hooksPath` locally. Lefthook installs into whichever hooks directory Git resolves, so a contributor who sets that option globally gets this repository's hook in their global directory, with any existing `pre-commit` renamed to `pre-commit.old`. Setting `core.hooksPath` locally in this repository avoids it. This is how lefthook behaves in every project that uses it.

### Verification

- `pnpm run lint-fix`, `pnpm tsc`, and `pnpm test packages/ariakit-scripts` (176 tests) pass.
- `pnpm exec lefthook validate` reports a valid config.
- A glob-skipped job does not stop the piped run: with no `package.json` staged, `clean` is skipped and the whole `lint` group still runs.
- An end to end `lefthook run pre-commit` against the real staged change confirmed that `clean` ran from the staged root `package.json`, that a nested file was linted and formatted, that the fixes were re-staged, and that an unstaged hunk stayed out of the staged content and was restored to the working tree.
- A throwaway pnpm project reproduced the install path in four states: a fresh repository installs the hook and a real `git commit` runs it, `CI=true` installs nothing, a directory without `.git` still exits 0, and a checkout carrying husky's `core.hooksPath` installs into `.husky/_` where a real `git commit` runs it.
- Long argument lists behave like lint-staged: 3001 staged files (610 KB of arguments) were split across 3 invocations rather than overflowing the command line.

## Review gate reassessment

<details>
<summary>Cycle 3: replace the install script with lefthook's documented install path</summary>

**Assessment**

Every finding-bearing cycle came from one place: how the Git hooks get installed, not from the hook configuration. `lefthook.yml` drew only comment-level findings, and the three findings against its semantics (`stage_fixed` without a file template, group level `{staged_files}` filtering, argument chunking) were all measured invalid.

Cycle 1 answered a real problem, `lefthook install` failing where husky returned early, with a 48 line `scripts/install-git-hooks.js` and a five test suite. Cycle 2 then spent four of its six findings maintaining that script: its test was vacuous under the `LEFTHOOK=0` that CI exports, and its `core.hooksPath` cleanup was destructive and uncovered. The machinery was generating its own findings, which is the signal this reassessment exists to catch.

The edge cases that drove it are cheap to weigh. A missing `.git`, which `ariakit react18` produces, is certain and breaks `pnpm test-react18`; lefthook's postinstall already exits 0 there. `CI` is certain; the postinstall already returns early. A stale husky `core.hooksPath` affects every existing checkout; lefthook writes the hook into that directory and it keeps working. A global `core.hooksPath` is rare, and lefthook's behavior there is the same in every project that uses it, so it belongs in release notes rather than in repository code.

**Decision**

Continue with lefthook's documented npm install path: `allowBuilds: lefthook: true` and no `prepare` script. pnpm 11 fails with `ERR_PNPM_IGNORED_BUILDS` when a dependency with a build script has no entry, so the entry is required regardless of its value, which makes this the smallest configuration that installs hooks at all.

Intentionally unsupported: automatically clearing husky's `core.hooksPath`, since lefthook's `--reset-hooks-path` also unsets the global option; the orphaned `.husky` directory, which the command above removes; and reverting auto fixes when a later job fails.

</details>

### CODE REVIEW

File: lefthook.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+pre-commit:
+  # `clean` rewrites the package exports that the type-aware linter resolves,
+  # so it has to finish first, and a failed job has to stop the ones after it.
+  piped: true
+  jobs:
+    - name: clean
+      glob: "**/package.json"
+      run: pnpm -r run --if-present clean
+      stage_fixed: true
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply lefthook.yml:16]: This is not an issue because `stage_fixed` re-stages only the job's filtered staged files, which is exactly what lint-staged did before this PR.

Measured on lefthook 2.1.10, with only `packages/a/package.json` staged while the job rewrites both workspaces:

- `packages/a/package.json` is re-staged with the rewrite.
- `packages/b/package.json` keeps the rewrite in the working tree and stays unstaged, with its own unstaged edit intact.

That is the intended behavior rather than a gap. Staging `b` would pull an unrelated workspace's modification into the contributor's commit, and the previous `"package.json": "pnpm -r run --if-present clean"` lint-staged entry scoped its `git add` the same way. So there is no behavior change here to cover with a cross-workspace test.

File: packages/ariakit-scripts/src/ci.ts [hunk 2]
```diff
@@ -106,8 +106,7 @@ const rootMainDependencies = new Set([
   "@types/react-dom",
   "@vitest/coverage-v8",
   "happy-dom",
-  "husky",
-  "lint-staged",
+  "lefthook",
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-scripts/src/ci.ts:109]: This is not an issue because the planner fails closed here instead of skipping `main`.

`addPackageJSONReasons` returns `false` when a changed root dependency is outside `rootMainDependencies`, and again when the changed fields are not dependency-only. Both hold for this PR, since it removes `husky` and `lint-staged` and also drops the `prepare` script. That return value does not narrow CI; it falls through to the full-CI path, which the existing "fails closed without dependency details and for CI infrastructure" test pins for `package.json`.

This PR's own run confirms it: every workflow was selected, including `Main / Lint`, `Main / Test`, and `Main / Test React 18`, alongside App, Docs, Performance, Plus, Release, and OG.

Adding `husky` and `lint-staged` back to the allowlist would have the opposite effect of the one suggested. It would let a root tooling migration take the narrow path, and both packages no longer exist anywhere in the repository.

File: pnpm-workspace.yaml [hunk 3]
```diff
@@ -11,6 +11,8 @@ allowBuilds:
   "@clerk/shared": false
   core-js: false
   esbuild: true
+  # Lefthook installs the Git hooks from its postinstall script.
+  lefthook: true
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply pnpm-workspace.yaml:15]: Not changing this in the current PR because the behavior belongs to lefthook's documented install path rather than to this repository, and every alternative measured worse here.

The finding is accurate: the postinstall runs `lefthook install -f`, which writes to whichever hooks directory Git resolves, so a globally configured `core.hooksPath` receives this repository's hook and an existing `pre-commit` is renamed to `pre-commit.old`. That is identical in every project that follows lefthook's npm installation guide, and 2.1.10 is the current release, so there is no version that changes it.

The options considered:

- A repository-owned install script that cleared only a local husky path. Removed as unnecessary machinery for a hook installer.
- `lefthook install --reset-hooks-path`. Measured to unset the contributor's **global** `core.hooksPath` outright, which is strictly more destructive than the behavior it avoids.
- `prepare: lefthook install || exit 0`, which refuses instead of overwriting. It protects the rare global setup, but leaves every existing checkout with no pre-commit hook until someone runs a manual `git config` command, so it regresses the common case to protect the uncommon one.

The behavior is recorded under Known behavior changes in the PR description, together with the workaround of setting `core.hooksPath` locally in this repository.

**Comment by diegohaz** [comment]
Addressed the three unresolved review threads. No code changes were needed, so the PR head is unchanged.

**Invalid**

- `lefthook.yml:16` ([reply](https://github.com/ariakit/ariakit/pull/7276#discussion_r3846063858)): `stage_fixed` re-stages only the job's filtered staged files, which is what the previous lint-staged entry did. Measured on lefthook 2.1.10 with one of two workspaces staged: the staged `package.json` is re-staged with the rewrite, and the unstaged one keeps the rewrite in the working tree but stays out of the commit with its own edit intact. Staging it would pull an unrelated workspace into the contributor's commit.
- `packages/ariakit-scripts/src/ci.ts:109` ([reply](https://github.com/ariakit/ariakit/pull/7276#discussion_r3846064501)): the planner fails closed rather than skipping `main`. `addPackageJSONReasons` returns `false` for a root dependency outside the allowlist, and again for changed fields that are not dependency-only, and that path selects full CI. This PR's own run selected every workflow, including `Main / Lint`, `Main / Test`, and `Main / Test React 18`. Re-adding `husky` and `lint-staged` to the allowlist would narrow CI for a root tooling migration instead.

**Declined**

- `pnpm-workspace.yaml:15` ([reply](https://github.com/ariakit/ariakit/pull/7276#discussion_r3846064731)): the report is accurate, but the behavior belongs to lefthook's documented install path and applies to every project that follows it, with no release that changes it. A repository-owned install script was removed as unnecessary machinery, `--reset-hooks-path` was measured to unset the contributor's global `core.hooksPath` outright, and `lefthook install || exit 0` leaves every existing checkout without hooks until a manual command. Recorded under Known behavior changes in the description.
